### ISSUE TITLE
Segmentation fault with a delete call inside chained prototyped sub calls

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12976]
Migrated from [rt.perl.org#118043](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=118043) (status was 'resolved')

Searchable as RT118043$

**Comment by p5pRT** [comment]
### From boyd.paul2@gmail.com

#### Created by boyd.paul2@gmail.com

This is a bug report for perl from boyd\.paul2@&#8203;gmail\.com\,
generated with the help of perlbug 1\.39 running under perl 5\.16\.3\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
I'm not sure how to explain this in words\, here's the code&#8203;:

sub a\(&&\) \{\};
sub b\(&\) \{\};
a \{\} b \{ delete $x \};

\[pboyd@&#8203;pboyd ~\]$ perl \-E 'sub a\(&&\) \{\}; sub b\(&\) \{\}; a \{\} b \{ delete $x \};'
delete argument is not a HASH or ARRAY element or slice at \-e line 1\.
Segmentation fault

I've tried this on v5\.12\.1\, v5\.16\.1 and v5\.16\.3\, perl segfaults on all of
them\.

It seems to be related to the prototypes\, specifically using sub \{\} make the
segfault go away\. Also just a single prototyped call \("a \{ delete $x \}"\)\,
doesn't segfault either\.

I ran into this with Try&#8203;::Tiny\, when I mistakenly did a "delete $foo" inside
the catch block\, something like this&#8203;:

perl \-E 'use Try&#8203;::Tiny; try \{\} catch \{ delete $x \}'

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.16.3:

Configured by nobody at Thu Apr 18 17:35:32 CEST 2013.

Summary of my perl5 (revision 5 version 16 subversion 3) configuration:

  Platform:
    osname=linux, osvers=3.8.7-1-arch, archname=i686-linux-thread-multi
    uname='linux flo32 3.8.7-1-arch #1 smp preempt sat apr 13 09:01:47 cest
2013 i686 gnulinux '
    config_args='-des -Dusethreads -Duseshrplib -Doptimize=-march=i686
-mtune=generic -O2 -pipe -fstack-protector --param=ssp-buffer-size=4
-Dprefix=/usr -Dvendorprefix=/usr -Dprivlib=/usr/share/perl5/core_perl
-Darchlib=/usr/lib/perl5/core_perl -Dsitelib=/usr/share/perl5/site_perl
-Dsitearch=/usr/lib/perl5/site_perl
-Dvendorlib=/usr/share/perl5/vendor_perl
-Dvendorarch=/usr/lib/perl5/vendor_perl -Dscriptdir=/usr/bin/core_perl
-Dsitescript=/usr/bin/site_perl -Dvendorscript=/usr/bin/vendor_perl
-Dinc_version_list=none -Dman1ext=1perl -Dman3ext=3perl -Dlddlflags=-shared
-Wl,-O1,--sort-common,--as-needed,-z,relro
-Dldflags=-Wl,-O1,--sort-common,--as-needed,-z,relro'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing
-pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE
-D_FILE_OFFSET_BITS=64',
    optimize='-march=i686 -mtune=generic -O2 -pipe -fstack-protector
--param=ssp-buffer-size=4',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe
-fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.8.0 20130411 (prerelease)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags ='-Wl,-O1,--sort-common,--as-needed,-z,relro
-fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
-lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=/lib/libc-2.17.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.17'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E
-Wl,-rpath,/usr/lib/perl5/core_perl/CORE'
    cccdlflags='-fPIC', lddlflags='-shared
-Wl,-O1,--sort-common,--as-needed,-z,relro -L/usr/local/lib
-fstack-protector'

Locally applied patches:



@INC for perl 5.16.3:
    /home/pboyd/perl5/lib/perl5/i686-linux-thread-multi
    /home/pboyd/perl5/lib/perl5/i686-linux-thread-multi
    /home/pboyd/perl5/lib/perl5
    /usr/lib/perl5/site_perl
    /usr/share/perl5/site_perl
    /usr/lib/perl5/vendor_perl
    /usr/share/perl5/vendor_perl
    /usr/lib/perl5/core_perl
    /usr/share/perl5/core_perl
    .


Environment for perl 5.16.3:
    HOME=/home/pboyd
    LANG=C
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)

PATH=/home/pboyd/bin:/home/pboyd/perl5/bin:/home/pboyd/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin:/sbin:/usr/bin/site_perl:/usr/bin/core_perl

PERL5LIB=/home/pboyd/perl5/lib/perl5/i686-linux-thread-multi:/home/pboyd/perl5/lib/perl5
    PERL_BADLANG (unset)
    PERL_LOCAL_LIB_ROOT=/home/pboyd/perl5
    PERL_MB_OPT=--install_base /home/pboyd/perl5
    PERL_MM_OPT=INSTALL_BASE=/home/pboyd/perl5
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Fri May 17 16&#8203;:52&#8203;:56 2013\, boyd\.paul2@&#8203;gmail\.com wrote&#8203;:
> This is a bug report for perl from boyd\.paul2@&#8203;gmail\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.16\.3\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> I'm not sure how to explain this in words\, here's the code&#8203;:
> 
> sub a\(&&\) \{\};
> sub b\(&\) \{\};
> a \{\} b \{ delete $x \};
> 
> \[pboyd@&#8203;pboyd ~\]$ perl \-E 'sub a\(&&\) \{\}; sub b\(&\) \{\}; a \{\} b \{ delete
> $x \};'
> delete argument is not a HASH or ARRAY element or slice at \-e line 1\.
> Segmentation fault
> 
> I've tried this on v5\.12\.1\, v5\.16\.1 and v5\.16\.3\, perl segfaults on all
> of
> them\.
> 
> It seems to be related to the prototypes\, specifically using sub \{\}
> make the
> segfault go away\. Also just a single prototyped call \("a \{ delete $x
> \}"\)\,
> doesn't segfault either\.
> 
> I ran into this with Try&#8203;::Tiny\, when I mistakenly did a "delete $foo"
> inside
> the catch block\, something like this&#8203;:
> 
> perl \-E 'use Try&#8203;::Tiny; try \{\} catch \{ delete $x \}'
> 

The segmentation fault \-\- or\, perhaps more precisely\, the \*report\* of a
segmentation fault \-\- is something specific to your operating
system/platform\.  I was able to get the same error message and notice of
segfault that you did when I ran it on Linux/i386\.

However\, when I ran it on Darwin/PPC\, I got only the error message \-\- no
notice of segfault\.

But\, IMO\, the segfault is irrelevant\.   Your code fails to compile&#8203;:

\#\#\#\#\#
$ cat 118043\-proto\.pl 
use 5\.010\_001;

sub a\(&&\) \{\};
sub b\(&\) \{\};
a \{\} b \{ delete $x \};

$ perl \-c 118043\-proto\.pl
delete argument is not a HASH or ARRAY element or slice at
118043\-proto\.pl line 5\.
\#\#\#\#\#

This is not a bug in Perl\.

Closing ticket\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From @tonycoz

On Fri May 17 17&#8203;:06&#8203;:38 2013\, jkeenan wrote&#8203;:
> The segmentation fault \-\- or\, perhaps more precisely\, the \*report\* of a
> segmentation fault \-\- is something specific to your operating
> system/platform\.  I was able to get the same error message and notice of
> segfault that you did when I ran it on Linux/i386\.
> 
> However\, when I ran it on Darwin/PPC\, I got only the error message \-\- no
> notice of segfault\.
> 
> But\, IMO\, the segfault is irrelevant\.   Your code fails to compile&#8203;:

A segmentation fault when failing to compile perl code is relevant\, and
is still a bug in perl\.

That said\, this has been fixed in blead\, in&#8203;:

commit 8be227ab5eaa23f2d21fd15f70190e494496dcbe
Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
Date&#8203;:   Sat Jun 23 09&#8203;:54&#8203;:31 2012 \-0700

&nbsp;    CV\-based slab allocation for ops
&nbsp;    
&nbsp;    This addresses bugs \#111462 and \#112312 and part of \#107000\.
&nbsp;    
&nbsp;    When a longjmp occurs during lexing\, parsing or compilation\, any ops
&nbsp;    in C autos that are not referenced anywhere are leaked\.
\.\.\.

This is unlikely to be backported to maint\-5\.16\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'rejected' to 'resolved'
